### ISSUE TITLE
训练模型无效果，test输出全为背景

### ISSUE THREAD

**Issue by Geslalala** [XSforAI/SUnet#1]
首先感谢你在本工作做出的贡献。以下是我的问题：1.我在复现本论文时完全按照指定数据集、指定的数据预处理方式以及指定预训练模型，然而训练出的model几乎没有任何效果，导致test输出全为0。我想问一下可能是哪里出现了问题？ 2.args.is_pretrain = False时，为什么依然需要调用预训练模型pvt_v2_b1.pth的参数？ 期待您的回复，谢谢！

**Comment by XSforAI** [comment]
> 首先感谢你在本工作做出的贡献。以下是我的问题：1.我在复现本论文时完全按照指定数据集、指定的数据预处理方式以及指定预训练模型，然而训练出的model几乎没有任何效果，导致test输出全为0。我想问一下可能是哪里出现了问题？ 2.args.is_pretrain = False时，为什么依然需要调用预训练模型pvt_v2_b1.pth的参数？ 期待您的回复，谢谢！

1.你使用的是哪个数据集呢？是否可以提供训练过程中的Loss迭代过程？
2.如果你不想使用预训练权重你可以注释掉以下语句或者依赖args.is_pretrain决定是否加载
net.load_from('./pretrain_pth/pvt_v2_b1.pth')

**Comment by Geslalala** [comment]
> > 首先感谢你在本工作做出的贡献。以下是我的问题：1.我在复现本论文时完全按照指定数据集、指定的数据预处理方式以及指定预训练模型，然而训练出的model几乎没有任何效果，导致test输出全为0。我想问一下可能是哪里出现了问题？ 2.args.is_pretrain = False时，为什么依然需要调用预训练模型pvt_v2_b1.pth的参数？ 期待您的回复，谢谢！
> 
> 1.你使用的是哪个数据集呢？是否可以提供训练过程中的Loss迭代过程？ 2.如果你不想使用预训练权重你可以注释掉以下语句或者依赖args.is_pretrain决定是否加载 net.load_from('./pretrain_pth/pvt_v2_b1.pth')

感谢您如此快速的回复。
1.我使用的是Synapse数据集，这是我在模型训练时的log
[training_log.txt](https://github.com/user-attachments/files/19362830/training_log.txt)
2.args.is_pretrain = False时，net.load_from('./pretrain_pth/pvt_v2_b1.pth')依然会运行。按照你的解释，我应该同时进行False设置和代码段的删除。

**Comment by XSforAI** [comment]
> > > 首先感谢你在本工作做出的贡献。以下是我的问题：1.我在复现本论文时完全按照指定数据集、指定的数据预处理方式以及指定预训练模型，然而训练出的model几乎没有任何效果，导致test输出全为0。我想问一下可能是哪里出现了问题？ 2.args.is_pretrain = False时，为什么依然需要调用预训练模型pvt_v2_b1.pth的参数？ 期待您的回复，谢谢！
> > 
> > 
> > 1.你使用的是哪个数据集呢？是否可以提供训练过程中的Loss迭代过程？ 2.如果你不想使用预训练权重你可以注释掉以下语句或者依赖args.is_pretrain决定是否加载 net.load_from('./pretrain_pth/pvt_v2_b1.pth')
> 
> 感谢您如此快速的回复。 1.我使用的是Synapse数据集，这是我在模型训练时的log [training_log.txt](https://github.com/user-attachments/files/19362830/training_log.txt) 2.args.is_pretrain = False时，net.load_from('./pretrain_pth/pvt_v2_b1.pth')依然会运行。按照你的解释，我应该同时进行False设置和代码段的删除。

1.按照你的loss，我们可以看到训练过程应该是没问题的，错误可能test脚本带来的。你可以参考https://github.com/SLDGroup/CASCADE的test。不过SUnet没有采用深度监督，要稍做修改。同时我们将会在一周内确认上传的代码是否存在问题。请稍后，但我们可以确认这个错误不是SUnet网络架构带来的，因为我们在我们新的MSDUNet一文中验证过。
同时我要再次确认，如下图的两个参数你是否修改为一致。

![Image](https://github.com/user-attachments/assets/939b7f07-73bc-40f8-8cc4-40521df45378)

2. 严格的来说你可以这么做：
if args.is_pretrain：
    net.load_from('./pretrain_pth/pvt_v2_b1.pth')
这样可以控制是否加载预训练模型。

**Comment by Geslalala** [comment]
> > > > 首先感谢你在本工作做出的贡献。以下是我的问题：1.我在复现本论文时完全按照指定数据集、指定的数据预处理方式以及指定预训练模型，然而训练出的model几乎没有任何效果，导致test输出全为0。我想问一下可能是哪里出现了问题？ 2.args.is_pretrain = False时，为什么依然需要调用预训练模型pvt_v2_b1.pth的参数？ 期待您的回复，谢谢！
> > > 
> > > 
> > > 1.你使用的是哪个数据集呢？是否可以提供训练过程中的Loss迭代过程？ 2.如果你不想使用预训练权重你可以注释掉以下语句或者依赖args.is_pretrain决定是否加载 net.load_from('./pretrain_pth/pvt_v2_b1.pth')
> > 
> > 
> > 感谢您如此快速的回复。 1.我使用的是Synapse数据集，这是我在模型训练时的log [training_log.txt](https://github.com/user-attachments/files/19362830/training_log.txt) 2.args.is_pretrain = False时，net.load_from('./pretrain_pth/pvt_v2_b1.pth')依然会运行。按照你的解释，我应该同时进行False设置和代码段的删除。
> 
> 1.按照你的loss，我们可以看到训练过程应该是没问题的，错误可能test脚本带来的。你可以参考[https://github.com/SLDGroup/CASCADE的test。不过SUnet没有采用深度监督，要稍做修改。同时我们将会在一周内确认上传的代码是否存在问题。请稍后，但我们可以确认这个错误不是SUnet网络架构带来的，因为我们在我们新的MSDUNet一文中验证过。](https://github.com/SLDGroup/CASCADE%E7%9A%84test%E3%80%82%E4%B8%8D%E8%BF%87SUnet%E6%B2%A1%E6%9C%89%E9%87%87%E7%94%A8%E6%B7%B1%E5%BA%A6%E7%9B%91%E7%9D%A3%EF%BC%8C%E8%A6%81%E7%A8%8D%E5%81%9A%E4%BF%AE%E6%94%B9%E3%80%82%E5%90%8C%E6%97%B6%E6%88%91%E4%BB%AC%E5%B0%86%E4%BC%9A%E5%9C%A8%E4%B8%80%E5%91%A8%E5%86%85%E7%A1%AE%E8%AE%A4%E4%B8%8A%E4%BC%A0%E7%9A%84%E4%BB%A3%E7%A0%81%E6%98%AF%E5%90%A6%E5%AD%98%E5%9C%A8%E9%97%AE%E9%A2%98%E3%80%82%E8%AF%B7%E7%A8%8D%E5%90%8E%EF%BC%8C%E4%BD%86%E6%88%91%E4%BB%AC%E5%8F%AF%E4%BB%A5%E7%A1%AE%E8%AE%A4%E8%BF%99%E4%B8%AA%E9%94%99%E8%AF%AF%E4%B8%8D%E6%98%AFSUnet%E7%BD%91%E7%BB%9C%E6%9E%B6%E6%9E%84%E5%B8%A6%E6%9D%A5%E7%9A%84%EF%BC%8C%E5%9B%A0%E4%B8%BA%E6%88%91%E4%BB%AC%E5%9C%A8%E6%88%91%E4%BB%AC%E6%96%B0%E7%9A%84MSDUNet%E4%B8%80%E6%96%87%E4%B8%AD%E9%AA%8C%E8%AF%81%E8%BF%87%E3%80%82) 同时我要再次确认，如下图的两个参数你是否修改为一致。
> 
> ![Image](https://github.com/user-attachments/assets/939b7f07-73bc-40f8-8cc4-40521df45378)
> 
> 2. 严格的来说你可以这么做：
>    if args.is_pretrain：
>    net.load_from('./pretrain_pth/pvt_v2_b1.pth')
>    这样可以控制是否加载预训练模型。


好的。1.图中的参数是一致的。我推测模型训练无效果的原因是：在utils.py文件中的def test_single_volume()，输入img和label正确，但是经过net后的p4很小，导致分类权值out在9个类别里，label=0最大，这可能是导致test中dice和HD95全为0的原因。

> def test_single_volume(image, label, net, classes, patch_size=[256, 256], test_save_path=None, case=None, z_spacing=1):
>     image, label = image.squeeze(0).cpu().detach().numpy(), label.squeeze(0).cpu().detach().numpy()
>     if len(image.shape) == 3:
>         prediction = np.zeros_like(label)
>         for ind in range(image.shape[0]):
>             slice = image[ind, :, :]
>             x, y = slice.shape[0], slice.shape[1]
>             if x != patch_size[0] or y != patch_size[1]:
>                 slice = zoom(slice, (patch_size[0] / x, patch_size[1] / y), order=3)  # previous using 0
>             input = torch.from_numpy(slice).unsqueeze(0).unsqueeze(0).float().cuda()
>             net.eval()
>             with torch.no_grad():
>                 p4 = net(input)
>                 outputs = p4
>                 out = torch.argmax(torch.softmax(outputs, dim=1), `dim=1).squeeze(0)

2.另外在test.py中，存在net = PVTUnet(num_classes=args.num_classes).cuda(0)导致test.py无法运行，我更改为了net = SUnet(num_classes=args.num_classes).cuda(0)。

再次感谢您的快速回复，我再调试一下代码以找出问题，后续我会继续跟进该项目。感谢！

**Comment by XSforAI** [comment]
> et(num_classes=args.num_classes).cuda(0)导致test.py无法运行，我更改为了net = SUnet(num_classes=args.num_classes).cuda(0)。

哈哈，真的抱歉，当时上线太过草率，我会在一周内整理代码的。

**Comment by XSforAI** [comment]
> > > > > 首先感谢你在本工作做出的贡献。以下是我的问题：1.我在复现本论文时完全按照指定数据集、指定的数据预处理方式以及指定预训练模型，然而训练出的model几乎没有任何效果，导致test输出全为0。我想问一下可能是哪里出现了问题？ 2.args.is_pretrain = False时，为什么依然需要调用预训练模型pvt_v2_b1.pth的参数？ 期待您的回复，谢谢！
> > > > 
> > > > 
> > > > 1.你使用的是哪个数据集呢？是否可以提供训练过程中的Loss迭代过程？ 2.如果你不想使用预训练权重你可以注释掉以下语句或者依赖args.is_pretrain决定是否加载 net.load_from('./pretrain_pth/pvt_v2_b1.pth')
> > > 
> > > 
> > > 感谢您如此快速的回复。 1.我使用的是Synapse数据集，这是我在模型训练时的log [training_log.txt](https://github.com/user-attachments/files/19362830/training_log.txt) 2.args.is_pretrain = False时，net.load_from('./pretrain_pth/pvt_v2_b1.pth')依然会运行。按照你的解释，我应该同时进行False设置和代码段的删除。
> > 
> > 
> > 1.按照你的loss，我们可以看到训练过程应该是没问题的，错误可能test脚本带来的。你可以参考[https://github.com/SLDGroup/CASCADE的test。不过SUnet没有采用深度监督，要稍做修改。同时我们将会在一周内确认上传的代码是否存在问题。请稍后，但我们可以确认这个错误不是SUnet网络架构带来的，因为我们在我们新的MSDUNet一文中验证过。](https://github.com/SLDGroup/CASCADE%E7%9A%84test%E3%80%82%E4%B8%8D%E8%BF%87SUnet%E6%B2%A1%E6%9C%89%E9%87%87%E7%94%A8%E6%B7%B1%E5%BA%A6%E7%9B%91%E7%9D%A3%EF%BC%8C%E8%A6%81%E7%A8%8D%E5%81%9A%E4%BF%AE%E6%94%B9%E3%80%82%E5%90%8C%E6%97%B6%E6%88%91%E4%BB%AC%E5%B0%86%E4%BC%9A%E5%9C%A8%E4%B8%80%E5%91%A8%E5%86%85%E7%A1%AE%E8%AE%A4%E4%B8%8A%E4%BC%A0%E7%9A%84%E4%BB%A3%E7%A0%81%E6%98%AF%E5%90%A6%E5%AD%98%E5%9C%A8%E9%97%AE%E9%A2%98%E3%80%82%E8%AF%B7%E7%A8%8D%E5%90%8E%EF%BC%8C%E4%BD%86%E6%88%91%E4%BB%AC%E5%8F%AF%E4%BB%A5%E7%A1%AE%E8%AE%A4%E8%BF%99%E4%B8%AA%E9%94%99%E8%AF%AF%E4%B8%8D%E6%98%AFSUnet%E7%BD%91%E7%BB%9C%E6%9E%B6%E6%9E%84%E5%B8%A6%E6%9D%A5%E7%9A%84%EF%BC%8C%E5%9B%A0%E4%B8%BA%E6%88%91%E4%BB%AC%E5%9C%A8%E6%88%91%E4%BB%AC%E6%96%B0%E7%9A%84MSDUNet%E4%B8%80%E6%96%87%E4%B8%AD%E9%AA%8C%E8%AF%81%E8%BF%87%E3%80%82) 同时我要再次确认，如下图的两个参数你是否修改为一致。
> > ![Image](https://github.com/user-attachments/assets/939b7f07-73bc-40f8-8cc4-40521df45378)
> > 
> > 2. 严格的来说你可以这么做：
> >    if args.is_pretrain：
> >    net.load_from('./pretrain_pth/pvt_v2_b1.pth')
> >    这样可以控制是否加载预训练模型。
> 
> 好的。1.图中的参数是一致的。我推测模型训练无效果的原因是：在utils.py文件中的def test_single_volume()，输入img和label正确，但是经过net后的p4很小，导致分类权值out在9个类别里，label=0最大，这可能是导致test中dice和HD95全为0的原因。
> 
> > def test_single_volume(image, label, net, classes, patch_size=[256, 256], test_save_path=None, case=None, z_spacing=1):
> > image, label = image.squeeze(0).cpu().detach().numpy(), label.squeeze(0).cpu().detach().numpy()
> > if len(image.shape) == 3:
> > prediction = np.zeros_like(label)
> > for ind in range(image.shape[0]):
> > slice = image[ind, :, :]
> > x, y = slice.shape[0], slice.shape[1]
> > if x != patch_size[0] or y != patch_size[1]:
> > slice = zoom(slice, (patch_size[0] / x, patch_size[1] / y), order=3)  # previous using 0
> > input = torch.from_numpy(slice).unsqueeze(0).unsqueeze(0).float().cuda()
> > net.eval()
> > with torch.no_grad():
> > p4 = net(input)
> > outputs = p4
> > out = torch.argmax(torch.softmax(outputs, dim=1), `dim=1).squeeze(0)
> 
> 2.另外在test.py中，存在net = PVTUnet(num_classes=args.num_classes).cuda(0)导致test.py无法运行，我更改为了net = SUnet(num_classes=args.num_classes).cuda(0)。
> 
> 再次感谢您的快速回复，我再调试一下代码以找出问题，后续我会继续跟进该项目。感谢！

让你久等，正确的训练脚本已上传，并增加了深度监督。如有需要请自取，给你带来的不便十分抱歉。
